### PR TITLE
add patience to the trainer for early stopping

### PR DESCRIPTION
**Pull request by baxtree**
This PR introduces a new parameter `patience` to `Trainer` for configuring the number of epochs to wait before early stopping the training if there is no consecutive improvement. It is inspired by [Keras's EarlyStopping ](https://keras.io/callbacks/#earlystopping).

### CODE REVIEW

**Comment by louisccc** [comment]
The idea of patience is great. but I also want the criteria (or just target) can be switched from loss to other metrics like filtered mean rank or hit-k ratio.

**Comment by baxtree** [comment]
Hi, @louisccc. Was thinking about that. Do you have a full list of metrics defined anywhere?

**Comment by louisccc** [comment]
The full list (like storing in array) isn't there. We should be having one, to be honest. 

Now we probably need to look into metriccalculator, all the stats are stored there. 

class MetricCalculator:
    '''
        MetricCalculator aims to 
        1) address all the statistic tasks.
        2) provide interfaces for querying results.

        MetricCalculator is expected to be used by "evaluation_process".
    '''
    def __init__(self, config):
        ...

        # (f)mr  : (filtered) mean rank
        # (f)mrr : (filtered) mean reciprocal rank
        # (f)hit : (filtered) hit-k ratio
        self.mr   = {}
        self.fmr  = {}
        self.mrr  = {}
        self.fmrr = {}
        self.hit  = {}
        self.fhit = {}
